runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java

```diff
@@ -212,24 +242,44 @@ public Timer timer(String timerId) {
     // on the output from "checkpoint".
     private @Nullable Instant residualWatermark;
     // A handle on the scheduled action to take a checkpoint.
-    private Future<?> scheduledCheckpoint;
+    private @Nullable Future<?> scheduledCheckpoint;
     private @Nullable Instant lastReportedWatermark;
 
     public ProcessContext(WindowedValue<InputT> element, TrackerT tracker) {
       fn.super();
       this.element = element;
       this.tracker = tracker;
+    }
 
-      this.scheduledCheckpoint =
-          executor.schedule(
-              () -> {
-                takeCheckpointNow();
-              },
-              maxDuration.getMillis(),
-              TimeUnit.MILLISECONDS);
+    @Override
+    public void onClaimed(PositionT position) {
+      checkState(
+          !hasClaimFailed,
+          "Must not call tryClaim() after it has previously returned false");
+      if (numClaimedBlocks == 0) {
+        // Claiming first block: can schedule the checkpoint now.
+        // We don't schedule it right away to prevent checkpointing before any blocks are claimed,
+        // in a state where no work has been done yet - because such a checkpoint is equivalent to
+        // the original restriction, i.e. pointless.
+        this.scheduledCheckpoint =
+            executor.schedule(
+                this::takeCheckpointNow, maxDuration.getMillis(), TimeUnit.MILLISECONDS);
```

Comment:
<!--new_thread; commit:c46da1bb17c0199bb58a8de052b59aedb875a013; resolved:0-->
This should be max((maxDuration - "time up to now"), 0) no ?

Reply:
I'm not sure. I think both "10 seconds of claimed work" and "10 seconds of total work" are valid options, but I'm slightly in favor of the former because it's less likely to lead to pathological behavior, e.g. imagine that opening a connection to Kafka consistently takes 10+ seconds due to network issues, then the first behavior will lead to repeatedly reading just 1 record from Kafka (compared to 0 before this PR...), but the second will provide 10 seconds of useful work.

Reply:
Sounds good.